### PR TITLE
`Development`: Remove outdated client dependencies

### PR DESCRIPTION
**Pull request by krusche**
Artemis uses a couple of outdated client dependencies which are not maintained anymore for several years.
This PR reduces the number of client dependencies from **146** to **107**.

It either inlines dependencies in a simplified version, removes them completely because they are obsolete, or replaces them with other libraries which are maintained more frequently. Here is a summary of the most important changes regarding the client production code:

- [x] Inline: **angular-fittext** (https://github.com/sollenne/angular-fittext.git)
- [x] Inline: **ng-starrating** (https://github.com/riteshgandhi/ng-star-rating)
- [x] Inline: **ngx-image-cropper** (https://github.com/Mawi137/ngx-image-cropper.git)
- [x] Inline: **ngx-treeview** (https://github.com/leovo2708/ngx-treeview.git)
- [x] Replace with angular material: **ng2-material-dropdown** (https://github.com/Gbuomprisco/ng2-material-dropdown5)  —> https://material.angular.io/components/chips/overview 
- [x] Replace with angular material: **ngx-chips** (https://github.com/Gbuomprisco/ngx-chips) —> https://material.angular.io/components/chips/overview 
- [x] Replace with angular material: **ng2-dnd** —> https://material.angular.io/cdk/drag-drop/overview
- [x] Replace with angular material: **ngx-clipboard** —> https://material.angular.io/cdk/clipboard/overview
- [x] Replace **jquery** with Angular Renderer and ViewChild
- [x] Finish the migration from `sinon` and `chai` to `Jest` in client tests
- [x] Remove all `Sinon` and `Chai` related dependencies
- [x] Use `dayjs/esm` instead of `dayjs`

Additional improvements (to be mentioned in release notes):
* `General`: Prevent showing exercises with two categories twice in the course administration
* `General`: Fix an edge case issue with statistics in the course administration (due to week overflows in the new year)
* `General`: Improve reuse of course categories during exercise creation
* `Development`: Reduce duplicated code in statistics related functionality
* `Development`: Update client and server dependencies
* `Development`: Update relevant files to Jhipster 7.5.0
* `Development`: Fix flaky server integration tests
* `Development`: Add new command to check to detect open handles in client tests
* `Development`: Fix memory leaks in client tests

**Please note:** There are some small remaining issues related to ratings and the usability of drag and drop quiz questions which we will fix in smaller follow-up PRs.

#### General

- [x] I tested **all** changes and their related features with **all** corresponding user types on a test server.
- [x] I chose a title conforming to the [naming conventions for pull requests](https://artemis-platform.readthedocs.io/en/latest/dev/guidelines/development-process.html#naming-conventions-for-github-pull-requests).
#### Client
- [x] I followed the [coding and design guidelines](https://docs.artemis.ase.in.tum.de/dev/guidelines/client/).
- [x] I added multiple integration tests (Jest) related to the features (with a high test coverage), while following the [test guidelines](https://docs.artemis.ase.in.tum.de/dev/guidelines/client-tests/).
- [x] I added `authorities` to all new routes and checked the course groups for displaying navigation elements (links, buttons).
- [x] I documented the TypeScript code using JSDoc style.

### Steps for Testing
We should focus on the following key areas while testing:

1. Tags
    a. Exercises Categories
    b. Server Administration --> User Management --> Groups
    c. Discussion --> Posts --> Tags
2. Drag and Drop (Quizzes)
3. File Browser in the Online Code Editor
4. Rating functionality

#### Code Review
- [x] Review 1
- [x] Review 2
#### Manual Tests
- [x] Test 1
- [x] Test 2

### Test Coverage
Due to the inline of some client dependencies, the coverage decreases. In follow-up PRs, we will add additional client tests to compensate this.

### Screenshots

image
image
image

### CODE REVIEW

**Review by willergeorg — COMMENTED** [review]
I tested the following things:

- Creation of exercise of every type with categories
- The file browser in the code editor
- Creation of a Post with categories
- The rating functionality for assessments for every type of exercise

During these cases, I encountered the following:

**For Exercises in general:**

- After creating the exercise, it is shown twice in the exercise list
- The feedback rating for assessments is not visible to me any more:

**For programming exercises:**

- Categories are not shown on the exercise detail page ("No categories" is displayed instead), but are shown in the edit view and to the student
- I did not encounter any problems with the file browser

File: src/main/webapp/app/exercises/shared/rating/star-rating/star-rating.component.ts [hunk 1]
```diff
@@ -0,0 +1,300 @@
...
+                // star on
+                StarRatingComponent.addCheckedStarClass(star);
+            } else {
+                // half star
+                if (hasDecimals) {
+                    StarRatingComponent.addHalfStarClass(star);
+                    hasDecimals = false;
+                }
+            }
+            i++;
```
Comment (b-fein) [inline src/main/webapp/app/exercises/shared/rating/star-rating/star-rating.component.ts:297]:
```suggestion
        this.stars.forEach((star: HTMLElement, i: number) => {
            star.className = '';
            this.applyColorStyle(star);
            StarRatingComponent.addDefaultClass(star);

            if (this.value > i) {
                // star on
                StarRatingComponent.addCheckedStarClass(star);
            } else {
                // half star
                if (hasDecimals) {
                    StarRatingComponent.addHalfStarClass(star);
                    hasDecimals = false;
                }
            }
```

There should be an overload on `forEach` to also be able to use the index directly.

**Review by b-fein — COMMENTED** [review]
Partial review of typescript production code.

At a glance on the test code I noticed that `expect(…).toEqual(…)` is used for simple values as well as objects. In the previous sinon/chai ⇒ jest conversions for simple values `toBe` was used instead. The guidelines for those cases are not specific, however.
I also noticed `toBeTrue()`/`toBeFalse()` instead of the explicitly preferred `toBe(true/false)`.

**Comment by krusche** [comment]
> Partial review of typescript production code.
> 
> At a glance on the test code I noticed that `expect(…).toEqual(…)` is used for simple values as well as objects. In the previous sinon/chai ⇒ jest conversions for simple values `toBe` was used instead. The guidelines for those cases are not specific, however. I also noticed `toBeTrue()`/`toBeFalse()` instead of the explicitly preferred `toBe(true/false)`.

During my search and replace approach (which was quite efficient), I noticed this as well. In my personal opinion, both ways in both cases that you are mentioning are fine. As long as the comparison works correctly and also detects wrong values, there is no big difference between `toBe` and `toEqual`. The effort to change all cases to the recommendation in the guidelines is not really worth in my opinion.

**Comment by willergeorg** [comment]
Dragging and dropping should work again:

**Pull request merged**